core-tools.markdown

```diff
@@ -0,0 +1,345 @@
+To enter the environment run the program `ghci`.
+
+```sh
+➜ ghci
+GHCi, version 9.0.1: https://www.haskell.org/ghc/  :? for help
+ghci> 
+```
+
+It provides an interactive prompt where Haskell expressions can be written and
+evaluated.
+
+For example:
+
+```sh
+ghci> 1 + 1
+2
+ghci> putStrLn "Hello, world!"
+Hello, world!
+```
+
+We can define new names:
+
+```sh
+ghci> double x = x + x
+ghci> double 2
+4
+```
+
+We can import Haskell source files using the `:load` command:
+
+```sh
+ghci> :load hello.hs
+[1 of 1] Compiling Main             ( hello.hs, interpreted )
+Ok, one module loaded.
+ghci> main
+Hello, Haskell!
+```
+
+As well as import library modules:
+
+```sh
+ghci> import Data.Bits
+ghci> shiftL 32 1
+64
+ghci> clearBit 33 0
+32
+```
+
+We can even ask what the type of an expression is using the `:type` command:
+
+```sh
+λ> :type putStrLn
+putStrLn :: String -> IO ()
+```
+
+To exit `ghci`, use the `:quit` command (or `:q` for short)
+
+```sh
+ghci> :quit
+Leaving GHCi.
+```
+
+A more thorough introduction to GHCi can be found in the [GHC user guide](https://downloads.haskell.org/~ghc/9.0.1/docs/html/users_guide/ghci.html).
+
+### Using external packages in ghci
+
+By default, ghci can only load and use packages that are [included with the GHC installation](https://downloads.haskell.org/ghc/latest/docs/html/users_guide/9.0.1-notes.html#included-libraries).
+
+However, users of [cabal-install](https://www.haskell.org/cabal) and [stack](http://haskellstack.org) can download and load external packages very easily using the following commands:
+
+Cabal-install:
+
+```sh
+cabal repl --build-depends random
+```
+
+Stack:
+
+```sh
+stack exec --package random -- ghci
+```
+
+And the modules of the relevant packages will be available for import:
+
+```sh
+GHCi, version 9.0.1: https://www.haskell.org/ghc/  :? for help
+λ> import System.Random
+λ> randomRIO (1, 10)
+1
+λ> randomRIO (1, 10)
+8
+λ> randomRIO (1, 10)
+5
+```
+
+Stack users can also use this feature with `runghc` and `ghc` by replacing `ghci` in the command above.
+
+Many more packages are waiting for you on [Hackage](https://hackage.haskell.org).
+
+## Compiling programs - Advanced usage
+
+In this section we'll cover more advanced usage available when compiling programs.
+Feel free to skip it!
+
+### Runtime options
+
+Compiling a Haskell program with the flag `-rtsopts` provides us with
+the ability to more finely control the Haskell runtime system when running the program.
+
+After compiling with `-rtsopts` with can pass flags to the Haskell runtime by
+listing them between `+RTS` and `-RTS` when invoking our Haskell program.
+For example, let's compile our `hello.hs` program with runtime options:
+
+```hs
+➜ ghc -rtsopts hello.hs 
+[1 of 1] Compiling Main             ( hello.hs, hello.o )
+Linking hello ...
+```
+
+And run it with `-sstderr` which will print some information
+about the runtime execution of the program: 
+
+```sh
+➜  ./hello +RTS -sstderr -RTS
```

Comment:
This is not a very compelling case for `-rtsopts` because you can get the same info with `./hello +RTS -s` without `-rtsopts` during compilation.

Reply:
You are right. I think according to feedback on this section it is better to remove the section altogether and link the GHC manual for more details.

Reply:
I don’t know: I think `-s` is pretty cool (in terms of all the info that it gives you for free). And I’m not big fun of sections like the last one on profiling where there’s just a reference and that’s it — it feels underwhelming. So it’s all about balance I guess: too much is overwhelming, too little is vice versa... It’s s your call in the end!

Reply:
I also think it's cool :) but I think in accordance with feedback from Rebecca and others it might be best to leave this kind of things out for now. We could always revisit this later and add it if we wish.